### PR TITLE
fix: add indicator when no Data on picker

### PR DESCRIPTION
**Pull request by gca-axelor**
RM#78121

### CODE REVIEW

**Comment by gca-axelor** [comment]
user screen 

```
/*
 * Axelor Business Solutions
 *
 * Copyright (C) 2024 Axelor (<http://axelor.com>).
 *
 * This program is free software: you can redistribute it and/or  modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, {useCallback, useEffect, useMemo} from 'react';
import DeviceInfo from 'react-native-device-info';
import {StyleSheet, Dimensions, View} from 'react-native';
import {
  Icon,
  ImageBubble,
  Picker,
  Screen,
  ScrollView,
  Text,
  useConfig,
  useTheme,
  useThemeColor,
} from '@axelor/aos-mobile-ui';
import {
  logout,
  useBinaryImageUri,
  useDispatch,
  useSelector,
  useTranslator,
} from '../../index';
import {fetchCompanies} from '../features/companySlice';
import {fetchLocalizations} from '../features/localizationSlice';
import {
  changeActiveCompany,
  updateActiveUser,
  fetchActiveUser,
} from '../features/userSlice';
import {PopupApplicationInformation} from '../../components';

const UserScreen = ({children}) => {
  const Theme = useTheme();
  const Colors = useThemeColor();
  const I18n = useTranslator();
  const formatImage = useBinaryImageUri();
  const dispatch = useDispatch();

  const {companyList} = useSelector(state => state.company);
  const {userId} = useSelector(state => state.auth);
  const {localizationList} = useSelector(state => state.localization);
  const {base: baseConfig} = useSelector(state => state.appConfig);
  const {loadingUser, user, isUser, canModifyCompany} = useSelector(
    state => state.user,
  );

  const {setFilterConfig, setVirtualKeyboardConfig, setNbDecimalDigitForQty} =
    useConfig();

  useEffect(() => {
    fetchUser();
    dispatch(fetchCompanies());
    dispatch(fetchLocalizations());
  }, [dispatch, fetchUser, userId]);

  useEffect(() => {
    if (baseConfig?.nbDecimalDigitForQty != null) {
      setNbDecimalDigitForQty(baseConfig?.nbDecimalDigitForQty);
    }
  }, [baseConfig, setNbDecimalDigitForQty]);

  useEffect(() => {
    const SMALL_SCREEN_HEIGHT = 500;

    DeviceInfo.getManufacturer().then(manufacturer =>
      setVirtualKeyboardConfig(manufacturer === 'Zebra Technologies'),
    );

    setFilterConfig(Dimensions.get('window').height > SMALL_SCREEN_HEIGHT);
  }, [setFilterConfig, setVirtualKeyboardConfig]);

  const styles = useMemo(() => {
    return getStyles(Colors);
  }, [Colors]);

  const updateActiveCompany = useCallback(
    company => {
      dispatch(
        changeActiveCompany({
          newCompany: {
            $version: company?.$version,
            code: company?.code,
            id: company?.id,
            name: company?.name,
          },
        }),
      );
    },
    [dispatch],
  );

  const handleChangeTheme = useCallback(
    newTheme => Theme.changeTheme(newTheme),
    [Theme],
  );

  const updateLanguage = useCallback(
    localization => {
      dispatch(
        updateActiveUser({
          id: user.id,
          localization: {id: localization},
          version: user.version,
        }),
      );
    },
    [dispatch, user],
  );

  const fetchUser = useCallback(() => {
    dispatch(fetchActiveUser(userId));
  }, [dispatch, userId]);

  return (
    <Screen style={styles.container}>
      <ScrollView
        refresh={{loading: loadingUser, fetcher: fetchUser}}
        style={styles.scroll}>
        <View style={styles.alignContainer}>
          <ImageBubble
            style={styles.imageIcon}
            imageSize={Dimensions.get('window').width * 0.3}
            defaultIconSize={60}
            source={formatImage(
              user?.id,
              user?.version,
              'com.axelor.auth.db.User',
            )}
            listComponent={[
              null,
              <Icon
                style={styles.logOutIcon}
                name="power"
                color={Colors.primaryColor.background}
                size={Dimensions.get('window').width * 0.07}
                touchable={true}
                onPress={() => dispatch(logout())}
              />,
            ]}
          />
          <Text style={styles.textUser}>{user.name}</Text>
          {baseConfig?.enableMultiCompany && (
            <Picker
              title={I18n.t('User_ActiveCompany')}
              listItems={companyList}
              defaultValue={user?.activeCompany}
              labelField="name"
              valueField="id"
              onValueChange={updateActiveCompany}
              isValueItem={true}
              readonly={!canModifyCompany}
            />
          )}
          {children}
          {localizationList?.length > 1 && (
            <Picker
              title={I18n.t('User_Language')}
              defaultValue={user.localization?.id}
              listItems={localizationList}
              labelField="name"
              valueField="id"
              onValueChange={updateLanguage}
              emptyValue={false}
            />
          )}
          {!Theme.isColorBlind && Theme.themes?.length !== 1 && (
            <>
              <Picker
                title={I18n.t('User_Theme')}
                defaultValue={Theme.activeTheme?.key}
                listItems={[]}
                labelField="name"
                valueField="key"
                onValueChange={handleChangeTheme}
                emptyValue={false}
              />
              <Picker
                title={I18n.t('User_Theme')}
                defaultValue={Theme.activeTheme?.key}
                listItems={Theme.themes}
                labelField="name"
                valueField="key"
                onValueChange={handleChangeTheme}
                emptyValue={false}
              />
            </>
          )}
          <PopupApplicationInformation
            visible={!isUser && !loadingUser}
            textKey={'User_NoAppUser'}
            onRefresh={fetchUser}
          />
        </View>
      </ScrollView>
    </Screen>
  );
};

const getStyles = Colors =>
  StyleSheet.create({
    alignContainer: {
      alignItems: 'center',
      paddingBottom: 150,
    },
    scroll: {
      height: null,
    },
    container: {
      justifyContent: 'center',
    },
    imageIcon: {
      backgroundColor: Colors.backgroundColor,
    },
    logOutIcon: {
      backgroundColor: Colors.backgroundColor,
      borderRadius: Dimensions.get('window').width * 0.1,
      width: Dimensions.get('window').width * 0.1,
      height: Dimensions.get('window').width * 0.1,
      elevation: 5,
      shadowOpacity: 0.5,
      shadowColor: Colors.secondaryColor.background,
      shadowOffset: {width: 0, height: 0},
    },
    textUser: {
      fontSize: 16,
      fontWeight: 'bold',
      marginBottom: 30,
      marginTop: 0,
    },
  });

export default UserScreen;

```

File: packages/ui/src/components/molecules/SelectionContainer/SelectionContainer.tsx [hunk 1]
```diff
@@ -153,6 +157,21 @@ const SelectionContainer = ({
     [keyField, selectedItem],
   );
 
+  const renderEmptyState = useCallback(() => {
+    const _title = title.toLowerCase();
+
+    const message =
+      translator != null
+        ? translator({key: 'Base_NoDataContainer', values: {_title}})
```
Comment (vhu-axelor) [inline packages/ui/src/components/molecules/SelectionContainer/SelectionContainer.tsx:165]: Not working: `translator('Base_NoDataContainer', {title: _title})`

File: packages/ui/src/components/molecules/SelectionContainer/SelectionContainer.tsx [hunk 2]
```diff
@@ -122,6 +122,8 @@ interface SelectionContainerProps {
   isPicker?: boolean;
   selectedItem?: any[];
   readonly?: boolean;
+  translator?: ({key, values}) => string;
```
Comment (vhu-axelor) [inline packages/ui/src/components/molecules/SelectionContainer/SelectionContainer.tsx:125]: Change type: `translator?: (key: string, values: Object) => string;`

File: packages/ui/src/components/organisms/MultiValuePicker/MultiValuePicker.tsx [hunk 3]
```diff
@@ -47,6 +47,7 @@ interface MultiValuePickerProps {
   listItems: Item[];
   readonly?: boolean;
   required?: boolean;
+  translator?: ({key, values}) => string;
```
Comment (vhu-axelor) [inline packages/ui/src/components/organisms/MultiValuePicker/MultiValuePicker.tsx:50]: Change type: `translator?: (key: string, values: Object) => string;`

File: packages/ui/src/components/organisms/Picker/Picker.tsx [hunk 4]
```diff
@@ -45,6 +45,7 @@ interface PickerProps {
   readonly?: boolean;
   required?: boolean;
   isScrollViewContainer?: boolean;
+  translator?: ({key, values}) => string;
```
Comment (vhu-axelor) [inline packages/ui/src/components/organisms/Picker/Picker.tsx:48]: Change type: `translator?: (key: string, values: Object) => string;`

File: packages/core/src/i18n/translations/en.json [hunk 5]
```diff
@@ -4,6 +4,7 @@
   "Base_OK": "Ok",
   "Base_NoMoreItems": "No more items.",
   "Base_NoData": "No data.",
+  "Base_NoDataContainer": "No {{title}} available",
```
Comment (vhu-axelor) [inline packages/core/src/i18n/translations/en.json:7]: I think "Base_NoDataPicker" is clearer

File: packages/core/src/i18n/translations/fr.json [hunk 6]
```diff
@@ -4,6 +4,7 @@
   "Base_OK": "Ok",
   "Base_NoMoreItems": "Fin de la liste.",
   "Base_NoData": "Aucune donnée.",
+  "Base_NoDataContainer": "Aucun {{title}} disponible",
```
Comment (vhu-axelor) [inline packages/core/src/i18n/translations/fr.json:7]: I think "Base_NoDataPicker" is clearer

File: changelogs/unreleased/78121.json [hunk 7]
```diff
@@ -0,0 +1,5 @@
+{
+  "title": "SelectionContainer: add indicator when no data",
```
Comment (vhu-axelor) [inline changelogs/unreleased/78121.json:2]: "Picker: add indicator when no data"

File: packages/ui/__tests__/components/molecules/SelectionContainer.test.js
Comment (vhu-axelor) [inline packages/ui/__tests__/components/molecules/SelectionContainer.test.js:1]: Change translator type function.  Change the two first return translations (to the french ones for example) to test if it really use the return translation or the default one when translator function is null

File: packages/ui/src/components/molecules/SelectionContainer/SelectionContainer.tsx [hunk 8]
```diff
@@ -153,6 +157,21 @@ const SelectionContainer = ({
     [keyField, selectedItem],
   );
 
+  const renderEmptyState = useCallback(() => {
+    const _title = title.toLowerCase();
```
Comment (vhu-axelor) [inline packages/ui/src/components/molecules/SelectionContainer/SelectionContainer.tsx:161]: Don't forget that title can be null

File: packages/ui/src/components/molecules/SelectionContainer/SelectionContainer.tsx [hunk 9]
```diff
@@ -153,6 +157,21 @@ const SelectionContainer = ({
...
+
+    const message =
+      translator != null
+        ? translator({key: 'Base_NoDataContainer', values: {_title}})
+        : `No ${_title} available`;
+
+    return (
+      <View style={styles.emptyContainer}>
```
Comment (vhu-axelor) [inline packages/ui/src/components/molecules/SelectionContainer/SelectionContainer.tsx:169]: `style={[styles.flatListContainer, styles.emptyContainer]}`

File: packages/ui/src/components/molecules/SelectionContainer/SelectionContainer.tsx [hunk 10]
```diff
@@ -262,6 +281,15 @@ const getStyles = (
...
     },
+    emptyContainer: {
+      top: -4,
+      padding: 10,
+      alignItems: 'center',
+      justifyContent: 'center',
+      borderColor: Colors.secondaryColor.background,
+      borderWidth: 1,
+      borderRadius: 7,
+    },
```
Comment (vhu-axelor) [inline packages/ui/src/components/molecules/SelectionContainer/SelectionContainer.tsx:292]:
```
emptyContainer: {
      alignItems: 'center',
      justifyContent: 'center',
}
```

File: packages/ui/src/components/molecules/SelectionContainer/SelectionContainer.tsx
Comment (vhu-axelor) [inline packages/ui/src/components/molecules/SelectionContainer/SelectionContainer.tsx:1]: In styles useMemo change listLength to 1 when no data indicator need to be display

File: packages/ui/src/components/organisms/Picker/Picker.tsx
Comment (vhu-axelor) [inline packages/ui/src/components/organisms/Picker/Picker.tsx:1]: Manage `isScrollViewContainer` when no data indicator need to be display

**Review by vhu-axelor — CHANGES_REQUESTED** [review]

**Pull request merged**